doc/src/devdocs/ub.md

```diff
@@ -0,0 +1,50 @@
+free of undefined behavior.
+
+The presence of undefined behavior is modeled as part of julia's effect system using the `:noub` effect bit. See the documentation for `@assume_effects` for more information on querying the compiler's effect model or overriding it for specific situations (e.g. where a dynamic check precludes potential UB from every actually being reached).
+
+## List of sources of undefined behavior
+
+The following is a list of sources of undefined behavior,
+though it should currently not be considered exhaustive:
+
+- Replacement of `const` values. Note that in interactive mode the compiler will issue a warning for this and some care is taken to mimize impact as a user convenience, but the behavior is not defined.
+- Various modification of global state during precompile. Where possible, this is detected and yields and error, but detection is incomplete.
+- Incorrect implementation of a `Core.OptimizedGenerics` interface [1]
+- Any invocation of undefined behavior in FFI code (e.g. `ccall`, `llvmcall`) according to the semantics of the respective language
+- Incorrect signature types in `ccall` or `cfunction`, even if those signatures happen to yield the correct ABI on a particular platform
+- Incorrect use of annotations like `@inbounds`, `@assume_effects` in violation of their requirements [1]
+- Retention of pointers to GC-tracked objects outside of a `@GC.preserve` region
+- Memory modification of GC-tracked objects without appropriate write barriers from outside of julia (e.g. in native calls, debuggers, etc.)
+- Violations of the memory model using `unsafe` operations (e.g. `unsafe_load` of an invalid pointer, pointer provenance violations, etc)
+- Violations of TBAA guarantees (e.g. using `unsafe_wrap`)
+- Mutation of data promised to be immutable (e.g. in `Base.StringVector`)
+- Data races
+- Modification of julia-internal mutable state (e.g. task schedulers, data types, etc.)
+
+[1] Incorrect use here may be UB, even if not semantically executed, please see the specific documentation of the feature.
+
+## Implementation-defined behavior
+Some behavior is technically forbidden by the semantics of the language, but required in certain parts of the implementation and thus allowed as long as implementation-defined constraints are obeyed. Nevertheless, these constructs should be avoided when possible, as the implementation-defined constraints may not be stable across julia versions.
+
+- Construction of objects using `eval(Expr(:new))` rather than their constructors
+
+## Special cases explicitly NOT undefined behavior
+
+- As of Julia 1.11, access to undefined bits types is no longer undefined behavior. It is still allowed to return an arbitrary value of the bits type, but the value returned must be the same for every access and use thereof is not undefined behavior. In LLVM terminology, the value is `freeze undef`, not `undef`.
```

Comment:
I think this is referring to cases like `x = Ref{Int}(); println(x[])`, right? Does this imply that e.g. padding can be assumed to be constant too?

Moreover, not all bitpatterns are necessarily valid values of the bitstype - e.g. a type such as

```julia
struct MaskedByte
    byte::UInt8
    MaskedByte(b::UInt8) = new(b & 0x0f)
end
```

doesn't have the same valid bitpatterns as `UInt8`, so `Ref{MaskedByte}()[]` can't guarantee to produce an arbitrary (valid) value of that type without going through the constructor:

```julia
julia> reinterpret(UInt8, Ref{MaskedByte}()[])
0x70
```

It's infeasible to guarantee calling a constructor for such uses though, so maybe the values produced like that (if not undefined) should have their own name? This would of course also cover invalid values of `@enum`, which currently uses the term `invalid` for this:

```julia
julia> @enum Foo A=0x0 B C D

julia> reinterpret(Foo, Int32(0xff))
<invalid #255>::Foo = 255
```

Reply:
Accessing padding is UB, as mentioned in the other comment. Violating inner constructor constraints for bits types is currently not UB or even disallowed, although I would like to clamp down on that in the future (by adding a bit in the type that disallows this).